src/main/extractZipInplace.js

```diff
@@ -0,0 +1,64 @@
+import fs from 'fs';
+import path from 'path';
+
+import yauzl from 'yauzl';
+
+import { getLogger } from '../logger';
+
+const logger = getLogger(__filename.split('/').slice(-1)[0]);
+
+/** Extract a zip archive to a directory with the same basename.
+ *
+ * Handle an arbitrary depth of files and folders within the archive,
+ * mirroring that structure in the output directory.
+ *
+ * @param {string} zipFilePath - path to the local zipfile.
+ * @returns { Promise } resolves true when done extracting all contents.
+ */
+export default function extractZipInplace(zipFilePath) {
+  return new Promise((resolve) => {
+    const extractToDir = path.dirname(zipFilePath);
+    logger.info(`extracting ${zipFilePath}`);
+    // lazyEntries allows explicit calls of readEntry,
+    // which we need to because we need to setup dirs as we go.
+    const options = {
+      lazyEntries: true,
+    };
+    yauzl.open(zipFilePath, options, (error, zipfile) => {
+      if (error) throw error;
+      zipfile.on('entry', (entry) => {
+        const writePath = path.join(extractToDir, entry.fileName);
+        // if entry is a directory
+        if (/\/$/.test(entry.fileName)) {
+          fs.mkdir(writePath, (e) => {
+            if (e) {
+              if (e.code !== 'EEXIST') { throw e; }
+            }
+            zipfile.readEntry();
+          });
+        } else {
+          zipfile.openReadStream(entry, (err, readStream) => {
+            if (err) throw err;
+            readStream.on('end', () => {
+              zipfile.readEntry();
+            });
+            // Sometimes an entry will be in a dir, where the
+            // dir itself was *not* an entry, therefore we still need
+            // to create the dir here.
+            fs.mkdir(path.dirname(writePath), (e) => {
```

Comment:
Just going to see if I've got a handle on what `fs.mkdir` is doing here.
`writePath` could be a file name in a dir structure that hasn't been made yet, so `mkdir` will make that directory chain if it is NOT already there and then go ahead and write the file.

Reply:
Oh but there is a bug here because `mkdir` does not make all the directories up the chain and there's no guarantee they all exist.